### PR TITLE
chore(kubelet): more maintainable kubeDeps conditions

### PR DESCRIPTION
**Pull request by jackfrancis**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This PR advocates that we trade a little bit of source code "elegance" for a more human-expressible, verbose implementation in order to promote easier maintenance going forward, and to avoid edge cases.

Specifically, overloading mutuallly exclusive nil conditional mutation with the existing switch-case implementation is a bit mind-bendy to grok at first, and doesn't track with the surrounding "if kubeDeps.<some kubeDeps property> == nil" expressions for all the other properties. Taken literally, the existing implementation declares:

- if either one of `kubeDeps.KubeClient` or `kubeDeps.EventClient` or `kubeDeps.HeartbeatClient` is nil, then overwrite *all* of them

That may make practical sense right now, as perhaps in practice those kubeDeps properties are tightly coupled in the above way. But because this code flow is only thinly wrapped inside of an exported `Run()` function that accepts any `kubeDeps *kubelet.Dependencies`, it's possible someone will pass in a `kubeDeps` with only one of the above properties set to nil, and not expect the other two to be overwritten.

This PR will change the above behaviors in the following way: kubelet will now only overwrite any of the three above `kubeDeps` properties _if they are actually nil_, and each will be evaluated discretely for nil. If that is actually behaviorally regressive, then I think that's probably a concern, so let's call that out and see why folks are actually expecting non-nil `kubeDeps` properties to be entirely overwritten during kubelet runtime instantiation.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by ehashman — COMMENTED** [review]
/priority backlog
/kind cleanup
/triage accepted
/release-note-none

**Comment by jackfrancis** [comment]
@klueska @mrunalp can you kindly take a look at this PR? thanks!

**Review by mrunalp — APPROVED** [review]

File: cmd/kubelet/app/server.go [hunk 1]
```diff
@@ -592,50 +592,57 @@ func run(ctx context.Context, s *options.KubeletServer, kubeDeps *kubelet.Depend
...
 			return errors.New("closeAllConns must be a valid function other than nil")
 		}
-		kubeDeps.OnHeartbeatFailure = closeAllConns
 
-		kubeDeps.KubeClient, err = clientset.NewForConfig(clientConfig)
-		if err != nil {
-			return fmt.Errorf("failed to initialize kubelet client: %w", err)
-		}
+		if kubeDeps.KubeClient == nil {
+			kubeDeps.OnHeartbeatFailure = closeAllConns
```
Comment (liggitt) [inline cmd/kubelet/app/server.go:612]: why is this inside the `if kubeDeps.KubeClient == nil` block?

Comment (liggitt) [reply cmd/kubelet/app/server.go:612]: /hold

this looks like a potential bug

Comment (jackfrancis) [reply cmd/kubelet/app/server.go]: @liggitt is right, this override as currently implemented is general when either `kubeDeps.KubeClient`, `kubeDeps.EventClient`, _or_ `kubeDeps.HeartbeatClient` is nil, it shouldn't be overriden *only* in the `kubeDeps.KubeClient == nil` scenario.

Fixed.

Thanks for catching this!

Comment (jackfrancis) [reply cmd/kubelet/app/server.go]: @klueska @mrunalp could you take another look at this (and thanks for your stamina!)

File: cmd/kubelet/app/server.go [hunk 2]
```diff
@@ -563,33 +563,39 @@ func run(ctx context.Context, s *options.KubeletServer, kubeDeps *kubelet.Depend
 		}
 		kubeDeps.OnHeartbeatFailure = closeAllConns
 
-		kubeDeps.KubeClient, err = clientset.NewForConfig(clientConfig)
-		if err != nil {
-			return fmt.Errorf("failed to initialize kubelet client: %w", err)
+		if kubeDeps.KubeClient == nil {
```
[the remark this answers is not in the document]
Comment (aojea) [reply cmd/kubelet/app/server.go:566]: ahh, no, nevermind, is an ||

**Comment by aojea** [comment]
agree on the statement of the description, but this alternative, with a lot of nested ifs, at least for me, is more difficult to read than the previous.

nested ifs is something we should avoid if possible, right?

**Comment by liggitt** [comment]
> this alternative, with a lot of nested ifs, at least for me, is more difficult to read than the previous.

I have the same reaction

**Comment by jackfrancis** [comment]
@aojea @liggitt I agree that this is less readable

The key point is that the current implementation overwrites all three client objects if any one of them are nil without that outcome being clearly expressed in the parent `Run()` interface contract. If we are 100% O.K. with this implementation, then we should close this PR.

If we do actually think we'd benefit by changing the implementation, I can work on improving the code surface area itself.

**Comment by aojea** [comment]
I think that if you can come with a better solution you should do it, but nested-ifs puts a lot of load on the reader to remember in which branch you are at some point :smile:

**Comment by jackfrancis** [comment]
> If this code is doing things that we wouldn't expect to work with partially populated clients, and no existing callers are setting those clients, it might be clearer to enforce an invariant that all three client fields are nil or all three are populated.

The "and no existing callers are setting those clients" part is the one we probably can't definitively prove. It doesn't *seem* that anyone would be doing that and not noticing that one or two of their pre-configured clients are being completely overwritten by the kubelet defaults enforcement flow... but it's certainly possible that someone out there is doing that. If we actually *enforce* all nil or none nil then those callers will be broken. This is pretty academic, but, it's probably safest keeping things the way they are (and I can't convince myself we're tackling something that has tangible user value to make it worth our while to make this change).

I threw this PR together however many months ago because I thought it would yield an easy code maintenance win. Seems in fact it's neither easier nor a win!

**Comment by jackfrancis** [comment]
I'm going to close this PR, big ocean, lots of fish, etc...

**Pull request closed without merging**